File: leetcode/medium/131/Solution.java
import java.util.*;

class Solution {
    public boolean isPalindrome(String s) {
        int l = 0;
        int r = s.length() - 1;

        while (l < r) {
            if (s.charAt(l) != s.charAt(r))
                return false;
        }

        return true;
    }

    public void backTrack(String s, List<String> path, List<List<String>> palindromes) {
        if (s.equals("")) {
            palindromes.add(path);
            System.out.println("added palindromes: " + path);
        }

        System.out.println("Current s: " + s);

        for (int i = 0; i < s.length(); i++) {
            String subString = s.substring(i, i + 1);
            System.out.println("current substring: " + subString);
            if (isPalindrome(subString)) {
                System.out.println("substring is palindrome");
                path.add(subString);
                System.out.println("substring added to path: " + path);
                backTrack(s.substring(i + 1, s.length()), path, palindromes);
                path.remove(path.size() - 1);
                System.out.println("substring removed from path: " + path);
            }
        }
    }

    public List<List<String>> partition(String s) {
        List<List<String>> palindromes = new ArrayList<>();
        List<String> path = new ArrayList<>();

        backTrack(s, path, palindromes);

        return palindromes;
    }

    public static void main(String[] args) {
        System.out.println(new Solution().partition("aab"));
    }
}

File: leetcode/medium/131/solution.py
from typing import List


class Solution:
    def partition(self, s: str) -> List[List[str]]:
        palindromes = []

        self.backtrack(s, [], palindromes)

        return palindromes

    def backtrack(self, s, path, palindromes):
        if s == "":
            palindromes.append(path)
            print("added palindromes: ", palindromes)

        print("current s: ", s)

        for i in range(len(s)):
            sub_str = s[: i + 1]
            print("current substr: ", sub_str)
            if self.is_palindrome(sub_str):
                # path.append(sub_str)
                # print("currently added path: ", path)
                # self.backtrack(s[i + 1 :], path, palindromes)
                self.backtrack(s[i + 1 :], path + [sub_str], palindromes)
                # path.pop(len(path) - 1)
                # print("return to previous state: ", path)

    # Helper method
    def is_palindrome(self, s) -> bool:
        return s == s[::-1]


if __name__ == "__main__":
    s = "aab"
    print(Solution().partition(s))
